ts/nni_manager/core/tuner_command_channel/websocket_channel.ts

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  The IPC channel between NNI manager and tuner.
+ *
+ *  TODO:
+ *   1. Merge with environment service's WebSocket channel.
+ *   2. Split import data command to avoid extremely long message.
+ *   3. Refactor message format.
+ **/
+
+import assert from 'assert/strict';
+import { EventEmitter } from 'events';
+
+import { Deferred } from 'ts-deferred';
+import type WebSocket from 'ws';
+
+import { Logger, getLogger } from 'common/log';
+
+const logger: Logger = getLogger('tuner_command_channel.WebSocketChannel');
+
+export interface WebSocketChannel {
+    init(): Promise<void>;
+    shutdown(): Promise<void>;
+    sendCommand(command: string): void;  // maybe this should return Promise<void>
+    onCommand(callback: (command: string) => void): void;
+    onError(callback: (error: Error) => void): void;
+}
+
+/**
+ *  Get the singleton tuner command channel.
+ *  Remember to invoke ``await channel.init()`` before doing anything else.
+ **/
+export function getWebSocketChannel(): WebSocketChannel {
+    return channelSingleton;
+}
+
+/**
+ *  The callback to serve WebSocket connection request. Used by REST server module.
+ *  It should only be invoked once, or an error will be raised.
+ *
+ *  Type hint of express-ws is somewhat problematic. Don't want to waste time on it so use `any`.
+ **/
+export function serveWebSocket(ws: WebSocket): void {
+    channelSingleton.setWebSocket(ws);
+}
+
+class WebSocketChannelImpl {
+    private deferredInit: Deferred<void> = new Deferred<void>();
+    private emitter: EventEmitter = new EventEmitter();
```

Comment:
Who's listening this emitter? If it's only this object itself, why not use the callbacks directly?

Reply:
`EventEmitter` is used as a "callback container". It deals with multiple handlers on one event.
Without it, we need either to maintain callback sets, or to make assertions and add docs. I think it is simpler to use a `EventEmitter`.